### PR TITLE
Fix/get stackstrings on move

### PR DESCRIPTION
**Pull request by williballenthin**
creating PR so we can discuss code.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -19,7 +19,8 @@ def extract_strings(vw):
     decoding_functions_candidates = identify_decoding_functions(vw)
     decoded_strings = floss_main.decode_strings(vw, function_index, decoding_functions_candidates)
     selected_functions = floss_main.select_functions(vw, None)
-    decoded_stackstrings = stackstrings.extract_stackstrings(vw, selected_functions)
+    bb_ends = stackstrings.get_basic_block_end_index(vw)
+    decoded_stackstrings = stackstrings.extract_stackstrings(vw, selected_functions, bb_ends)
```
Comment (williballenthin) [inline tests/conftest.py]: i think we can move the generation of the ends index into the `extract_stackstrings` function.

we don't reuse the index, so there's no benefit of keeping around.
also, as a user of the library, its just another thing you have to remember to do (construct the index and provide it as a parameter).

File: floss/stackstrings.py [hunk 2]
```diff
@@ -57,10 +64,33 @@ def apicall(self, emu, op, pc, api, argv):
         stack_buf = emu.readMemory(stack_top, stack_size)
         self.ctxs.append(CallContext(op.va, stack_top, stack_bottom, stack_buf))
 
-
-def extract_call_contexts(vw, fva):
+    def posthook(self, emu, op, endpc):
+        self.get_context_via_mov_heuristic(emu, op, endpc)
+
+    def get_context_via_mov_heuristic(self, emu, op, endpc):
```
Comment (williballenthin) [inline floss/stackstrings.py]: this doesn't return a context, so we should probably name it differently (reserve `get_*` for fetching some value). maybe `update_stack_heuristics()`.

Comment (williballenthin) [inline floss/stackstrings.py]: nvm, it does compute the context, though doesn't return it.

still recommend renaming it.

File: floss/stackstrings.py [hunk 3]
```diff
@@ -57,10 +64,33 @@ def apicall(self, emu, op, pc, api, argv):
...
+        """ Extract contexts at end of a basic block (bb) if bb contains enough movs to stack memory. """
+        # TODO check number of written bytes?
+        # count movs, shortcut if this basic block has enough writes to trigger context extraction already
+        if self.mov_count < MIN_NUMBER_OF_MOVS and self.is_stack_mov(emu, op):
```
Comment (williballenthin) [inline floss/stackstrings.py]: i like the short-circuit here, where the `is_stack_mov()` is not called if the threshold is already hit. 

this means that the `self.mov_count` number will not be perfectly accurate all the time (might be lower than the actual count). but i think this is ok, with a bit of documentation.

File: floss/stackstrings.py [hunk 4]
```diff
@@ -140,26 +188,30 @@ def extract_stackstrings(vw, selected_functions):
     :rtype: Generator[StackString]
     '''
     logger.debug('extracting stackstrings from %d functions', len(selected_functions))
+    bb_ends = get_basic_block_ends(vw)
```
Comment (mr-tz) [inline floss/stackstrings.py]: good call moving this here

File: floss/stackstrings.py [hunk 5]
```diff
@@ -57,10 +64,33 @@ def apicall(self, emu, op, pc, api, argv):
...
+        if self.mov_count < MIN_NUMBER_OF_MOVS and self.is_stack_mov(emu, op):
+            self.mov_count += 1
+        if endpc in self.bb_ends:
+            if self.mov_count >= MIN_NUMBER_OF_MOVS:
+                self.extract_context(emu, op)
+            # reset counter at end of basic block
+            self.mov_count = 0
+
+    def is_stack_mov(self, emu, op):
+        if op.mnem[:3] == "mov":
```
Comment (williballenthin) [inline floss/stackstrings.py]: using `op.mnem.startswith('mov')` is a micro-optimization that might not be worth doing. but should save a single allocation where the substring is sliced out and then compared.

File: floss/stackstrings.py [hunk 6]
```diff
@@ -57,10 +64,33 @@ def apicall(self, emu, op, pc, api, argv):
...
+        # count movs, shortcut if this basic block has enough writes to trigger context extraction already
+        if self.mov_count < MIN_NUMBER_OF_MOVS and self.is_stack_mov(emu, op):
+            self.mov_count += 1
+        if endpc in self.bb_ends:
+            if self.mov_count >= MIN_NUMBER_OF_MOVS:
+                self.extract_context(emu, op)
+            # reset counter at end of basic block
+            self.mov_count = 0
+
+    def is_stack_mov(self, emu, op):
```
Comment (williballenthin) [inline floss/stackstrings.py]: `emu` is not used in this function

Comment (williballenthin) [inline floss/stackstrings.py]: this function doesn't actually check to see if the destination is on the stack. do you think we should do this? i guess its pruned off elsewhere (in the ctx collection).

Comment (mr-tz) [inline floss/stackstrings.py]: For instructions like `movsd` checking the operand would require backtracking `EDI` etc. Instead I propose we count these movs and let the later extraction phases take care of it.

**Comment by mr-tz** [comment]
On the right this branch's code, on the left `self.check_mov_heuristics(emu, op, endpc)` in `posthook` is commented out.
```
FLOSS extracted 68 stackstrings				      |	FLOSS extracted 70 stackstrings
Finished execution after 12.437037 seconds		      |	Finished execution after 6.116257 seconds
FLOSS extracted 68 stackstrings				      |	FLOSS extracted 70 stackstrings
Finished execution after 12.601272 seconds		      |	Finished execution after 3.906816 seconds
FLOSS extracted 33 stackstrings					FLOSS extracted 33 stackstrings
Finished execution after 6.261168 seconds		      |	Finished execution after 2.102670 seconds
FLOSS extracted 33 stackstrings					FLOSS extracted 33 stackstrings
Finished execution after 8.648854 seconds		      |	Finished execution after 2.153056 seconds
FLOSS extracted 34 stackstrings					FLOSS extracted 34 stackstrings
Finished execution after 6.435290 seconds		      |	Finished execution after 2.103372 seconds
FLOSS extracted 7 stackstrings					FLOSS extracted 7 stackstrings
Finished execution after 1.500629 seconds		      |	Finished execution after 0.540202 seconds
FLOSS extracted 58 stackstrings					FLOSS extracted 58 stackstrings
Finished execution after 7.225434 seconds		      |	Finished execution after 2.104242 seconds
FLOSS extracted 33 stackstrings					FLOSS extracted 33 stackstrings
Finished execution after 6.456765 seconds		      |	Finished execution after 1.988064 seconds
FLOSS extracted 6 stackstrings					FLOSS extracted 6 stackstrings
Finished execution after 0.856545 seconds		      |	Finished execution after 0.290130 seconds
FLOSS extracted 4 stackstrings					FLOSS extracted 4 stackstrings
Finished execution after 0.379040 seconds		      |	Finished execution after 0.089091 seconds
FLOSS extracted 0 stackstrings					FLOSS extracted 0 stackstrings
Finished execution after 1.200029 seconds		      |	Finished execution after 0.301339 seconds
FLOSS extracted 5 stackstrings					FLOSS extracted 5 stackstrings
Finished execution after 3.667832 seconds		      |	Finished execution after 1.050123 seconds
FLOSS extracted 2 stackstrings					FLOSS extracted 2 stackstrings
Finished execution after 0.620089 seconds		      |	Finished execution after 0.128844 seconds
FLOSS extracted 0 stackstrings					FLOSS extracted 0 stackstrings
Finished execution after 1.656214 seconds		      |	Finished execution after 0.342097 seconds
FLOSS extracted 12 stackstrings					FLOSS extracted 12 stackstrings
Finished execution after 6.563190 seconds		      |	Finished execution after 1.596478 seconds
FLOSS extracted 0 stackstrings					FLOSS extracted 0 stackstrings
Finished execution after 1.257680 seconds		      |	Finished execution after 0.266993 seconds
FLOSS extracted 0 stackstrings					FLOSS extracted 0 stackstrings
Finished execution after 0.520367 seconds		      |	Finished execution after 0.114511 seconds
FLOSS extracted 4 stackstrings					FLOSS extracted 4 stackstrings
Finished execution after 0.767845 seconds		      |	Finished execution after 0.180638 seconds
FLOSS extracted 30 stackstrings					FLOSS extracted 30 stackstrings
Finished execution after 11.205189 seconds		      |	Finished execution after 13.020079 seconds
FLOSS extracted 26 stackstrings					FLOSS extracted 26 stackstrings
Finished execution after 6.165583 seconds		      |	Finished execution after 8.128161 seconds
FLOSS extracted 12 stackstrings					FLOSS extracted 12 stackstrings
Finished execution after 8.022916 seconds		      |	Finished execution after 8.327140 seconds
FLOSS extracted 23 stackstrings					FLOSS extracted 23 stackstrings
Finished execution after 7.388746 seconds		      |	Finished execution after 7.177046 seconds
```
Second run, same as above:
```
FLOSS extracted 68 stackstrings				      |	FLOSS extracted 70 stackstrings
Finished execution after 14.907161 seconds		      |	Finished execution after 11.001571 seconds
FLOSS extracted 68 stackstrings				      |	FLOSS extracted 70 stackstrings
Finished execution after 16.210848 seconds		      |	Finished execution after 12.956547 seconds
FLOSS extracted 33 stackstrings					FLOSS extracted 33 stackstrings
Finished execution after 6.852020 seconds		      |	Finished execution after 5.371959 seconds
FLOSS extracted 33 stackstrings					FLOSS extracted 33 stackstrings
Finished execution after 7.408496 seconds		      |	Finished execution after 7.835338 seconds
FLOSS extracted 34 stackstrings					FLOSS extracted 34 stackstrings
Finished execution after 4.616858 seconds		      |	Finished execution after 6.451429 seconds
FLOSS extracted 7 stackstrings					FLOSS extracted 7 stackstrings
Finished execution after 2.254808 seconds		      |	Finished execution after 1.668714 seconds
FLOSS extracted 58 stackstrings					FLOSS extracted 58 stackstrings
Finished execution after 6.931866 seconds		      |	Finished execution after 9.425688 seconds
FLOSS extracted 33 stackstrings					FLOSS extracted 33 stackstrings
Finished execution after 7.747519 seconds		      |	Finished execution after 5.270464 seconds
FLOSS extracted 6 stackstrings					FLOSS extracted 6 stackstrings
Finished execution after 1.219478 seconds		      |	Finished execution after 1.122699 seconds
FLOSS extracted 4 stackstrings					FLOSS extracted 4 stackstrings
Finished execution after 0.346893 seconds		      |	Finished execution after 0.355022 seconds
FLOSS extracted 0 stackstrings					FLOSS extracted 0 stackstrings
Finished execution after 0.690604 seconds		      |	Finished execution after 1.243299 seconds
FLOSS extracted 5 stackstrings					FLOSS extracted 5 stackstrings
Finished execution after 4.030465 seconds		      |	Finished execution after 3.981539 seconds
FLOSS extracted 2 stackstrings					FLOSS extracted 2 stackstrings
Finished execution after 0.283137 seconds		      |	Finished execution after 0.264660 seconds
FLOSS extracted 0 stackstrings					FLOSS extracted 0 stackstrings
Finished execution after 0.815422 seconds		      |	Finished execution after 0.559159 seconds
FLOSS extracted 12 stackstrings					FLOSS extracted 12 stackstrings
Finished execution after 6.256333 seconds		      |	Finished execution after 4.561625 seconds
FLOSS extracted 0 stackstrings					FLOSS extracted 0 stackstrings
Finished execution after 0.552247 seconds		      |	Finished execution after 0.724445 seconds
FLOSS extracted 0 stackstrings					FLOSS extracted 0 stackstrings
Finished execution after 0.224816 seconds		      |	Finished execution after 0.214960 seconds
FLOSS extracted 4 stackstrings					FLOSS extracted 4 stackstrings
Finished execution after 0.621071 seconds		      |	Finished execution after 0.306037 seconds
FLOSS extracted 30 stackstrings					FLOSS extracted 30 stackstrings
Finished execution after 13.112107 seconds		      |	Finished execution after 5.287577 seconds
FLOSS extracted 26 stackstrings					FLOSS extracted 26 stackstrings
Finished execution after 5.079447 seconds		      |	Finished execution after 2.245976 seconds
FLOSS extracted 12 stackstrings					FLOSS extracted 12 stackstrings
Finished execution after 8.054548 seconds		      |	Finished execution after 2.982275 seconds
FLOSS extracted 23 stackstrings					FLOSS extracted 23 stackstrings
Finished execution after 2.173112 seconds		      |	Finished execution after 3.917722 seconds
```

b8235be vs. this branch's code:
```
Finished execution after 16.627158 seconds		      |	Finished execution after 15.115348 seconds
Finished execution after 15.619825 seconds		      |	Finished execution after 9.052471 seconds
Finished execution after 3.752451 seconds		      |	Finished execution after 2.106583 seconds
Finished execution after 6.330209 seconds		      |	Finished execution after 2.096715 seconds
Finished execution after 7.664421 seconds		      |	Finished execution after 2.026900 seconds
Finished execution after 2.109611 seconds		      |	Finished execution after 0.567120 seconds
Finished execution after 7.316648 seconds		      |	Finished execution after 2.049298 seconds
Finished execution after 5.670480 seconds		      |	Finished execution after 8.677114 seconds
Finished execution after 1.179953 seconds		      |	Finished execution after 1.117290 seconds
Finished execution after 0.226506 seconds		      |	Finished execution after 0.193582 seconds
Finished execution after 1.013993 seconds		      |	Finished execution after 0.954240 seconds
Finished execution after 3.169766 seconds		      |	Finished execution after 1.697595 seconds
Finished execution after 0.223450 seconds		      |	Finished execution after 0.192907 seconds
Finished execution after 1.474137 seconds		      |	Finished execution after 0.358061 seconds
Finished execution after 2.344895 seconds		      |	Finished execution after 1.648813 seconds
Finished execution after 0.729818 seconds		      |	Finished execution after 0.283580 seconds
Finished execution after 0.207744 seconds		      |	Finished execution after 0.119851 seconds
Finished execution after 0.280792 seconds		      |	Finished execution after 0.169115 seconds
Finished execution after 9.012873 seconds		      |	Finished execution after 5.881229 seconds
Finished execution after 4.625523 seconds		      |	Finished execution after 2.250172 seconds
Finished execution after 8.576326 seconds		      |	Finished execution after 2.698020 seconds
Finished execution after 6.836436 seconds		      |	Finished execution after 1.835005 seconds
```

**Comment by williballenthin** [comment]
i don't really understand why this branch would be faster than b8235be. we are doing more work per instruction during stackstring extraction. right?

this has me worried, though it *seems* to be extracting the same amount of data.

**Comment by mr-tz** [comment]
Comparing https://github.com/fireeye/flare-floss/tree/separate_extract_context to this branch's code.
```
Finished execution after 12.099057 seconds		      |	Finished execution after 16.143504 seconds
Finished execution after 9.053385 seconds		      |	Finished execution after 9.404850 seconds
Finished execution after 3.749981 seconds		      |	Finished execution after 2.427076 seconds
Finished execution after 6.816325 seconds		      |	Finished execution after 1.918424 seconds
Finished execution after 7.430469 seconds		      |	Finished execution after 6.666122 seconds
Finished execution after 2.113906 seconds		      |	Finished execution after 1.318399 seconds
Finished execution after 10.317241 seconds		      |	Finished execution after 8.009269 seconds
Finished execution after 4.846394 seconds		      |	Finished execution after 3.019059 seconds
Finished execution after 1.175957 seconds		      |	Finished execution after 0.327849 seconds
Finished execution after 0.221410 seconds		      |	Finished execution after 0.091995 seconds
Finished execution after 0.515665 seconds		      |	Finished execution after 0.299789 seconds
Finished execution after 2.171128 seconds		      |	Finished execution after 1.023689 seconds
Finished execution after 0.200053 seconds		      |	Finished execution after 0.132735 seconds
Finished execution after 0.466327 seconds		      |	Finished execution after 0.325200 seconds
Finished execution after 1.406480 seconds		      |	Finished execution after 1.582044 seconds
Finished execution after 0.230163 seconds		      |	Finished execution after 0.271163 seconds
Finished execution after 0.095234 seconds		      |	Finished execution after 0.108953 seconds
Finished execution after 0.818630 seconds		      |	Finished execution after 0.176135 seconds
Finished execution after 12.667251 seconds		      |	Finished execution after 9.929335 seconds
Finished execution after 6.102029 seconds		      |	Finished execution after 8.580881 seconds
Finished execution after 8.236292 seconds		      |	Finished execution after 6.662575 seconds
Finished execution after 4.615740 seconds		      |	Finished execution after 3.195896 seconds
```

**Comment by williballenthin** [comment]
looks good. should we merge?

**Pull request merged**